In [ ]:
import pandas as pd
# dataset link: https://media.geeksforgeeks.org/wp-content/uploads/20240514105101/IMDB-Dataset.csv
df = pd.read_csv("../datasets/IMDB_movie_review_dataset.csv")
df.head()
len(df)

50000

In [50]:
import numpy as np

df["sentiment"] = np.where(df["sentiment"] == "positive", 1, 0)

In [51]:
review = df["review"]
sentiment = df["sentiment"]

In [52]:
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(review, sentiment, test_size=0.2, random_state=42)

In [63]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer()
X_train_transformed = vectorizer.fit_transform(X_train)
X_test_transformed = vectorizer.transform(X_test)


In [ ]:
X_train_transformed.shape
# print(X_test_transformed)

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 1361185 stored elements and shape (10000, 93003)>
  Coords	Values
  (0, 49)	1
  (0, 2703)	2
  (0, 3629)	1
  (0, 4094)	7
  (0, 4435)	1
  (0, 4670)	1
  (0, 5157)	1
  (0, 7196)	2
  (0, 8196)	1
  (0, 8836)	1
  (0, 9098)	1
  (0, 9106)	1
  (0, 10126)	1
  (0, 10134)	1
  (0, 10623)	1
  (0, 10636)	1
  (0, 11289)	1
  (0, 11669)	1
  (0, 12409)	1
  (0, 13333)	1
  (0, 14498)	1
  (0, 18762)	1
  (0, 20008)	1
  (0, 20014)	1
  (0, 21291)	1
  :	:
  (9999, 61184)	1
  (9999, 66852)	3
  (9999, 67783)	1
  (9999, 67944)	1
  (9999, 71588)	2
  (9999, 71774)	1
  (9999, 72636)	2
  (9999, 72724)	1
  (9999, 72739)	1
  (9999, 74272)	1
  (9999, 78668)	1
  (9999, 82276)	1
  (9999, 82298)	9
  (9999, 82372)	1
  (9999, 82416)	1
  (9999, 82525)	1
  (9999, 82568)	1
  (9999, 82641)	1
  (9999, 83250)	1
  (9999, 83481)	1
  (9999, 88265)	2
  (9999, 88411)	1
  (9999, 89569)	2
  (9999, 92055)	1
  (9999, 92267)	2


In [65]:
from sklearn.naive_bayes import MultinomialNB

multinomial_model = MultinomialNB()
multinomial_model.fit(X_train_transformed, Y_train)

MultinomialNB()

In [66]:
from sklearn.metrics import accuracy_score
# accuracy
print(f"Training Accuracy: ${accuracy_score(Y_train, multinomial_model.predict(X_train_transformed))}")
print(f"Testing Accuracy: ${accuracy_score(Y_test, multinomial_model.predict(X_test_transformed))}")

Training Accuracy: $0.890975
Testing Accuracy: $0.8488


In [56]:
review_input = input("Enter a review: ")
review_input = [review_input]
review_input = vectorizer.transform(review_input)
print("Positive" if multinomial_model.predict(review_input) == 1 else "Negative") 

Negative


In [57]:
# try a different model
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(X_train_transformed, Y_train)

print(f"Training Accuracy: ${accuracy_score(Y_train, model.predict(X_train_transformed))}")
print(f"Testing Accuracy: ${accuracy_score(Y_test, model.predict(X_test_transformed))}")

Training Accuracy: $0.95715
Testing Accuracy: $0.8921


f:\Work\hands-on-nlp\venv\Lib\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [58]:
review_input = input("Enter a review: ")
review_input = [review_input]
review_input = vectorizer.transform(review_input)
print("Positive" if model.predict(review_input) == 1 else "Negative") 

Positive


### Comment
Logistic Regression Model performs better. Accuracy of 0.89 whereas MultinomialNB gave 0.84

In [90]:
# Now perform nlp preprocessing
import nltk
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet
from nltk.tag import pos_tag
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS

In [81]:
nltk.download("punkt_tab",download_dir="f:\\Work\\hands-on-nlp\\venv\\nltk_data")
nltk.download("wordnet",download_dir="f:\\Work\\hands-on-nlp\\venv\\nltk_data")
nltk.download("omw-1.4",download_dir="f:\\Work\\hands-on-nlp\\venv\\nltk_data")
nltk.download("averaged_perceptron_tagger_eng",download_dir="f:\\Work\\hands-on-nlp\\venv\\nltk_data")

[nltk_data] Downloading package punkt_tab to f:\Work\hands-on-
[nltk_data]     nlp\venv\nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!
[nltk_data] Downloading package wordnet to f:\Work\hands-on-
[nltk_data]     nlp\venv\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to f:\Work\hands-on-
[nltk_data]     nlp\venv\nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     f:\Work\hands-on-nlp\venv\nltk_data...
[nltk_data]   Unzipping taggers\averaged_perceptron_tagger_eng.zip.


True

In [91]:
lemmatizer = WordNetLemmatizer()

def get_pos_given_word(word:str) :
    tag = pos_tag([word])[0][1][0].upper()
    tag_dict = {"J": wordnet.ADJ,
                "N": wordnet.NOUN,
                "V": wordnet.VERB,
                "R": wordnet.ADV}
    return tag_dict.get(tag, wordnet.NOUN)

def custom_tokenizer(text):
    tokens = word_tokenize(text)
    lemmatized_tokens = [lemmatizer.lemmatize(word.lower(), get_pos_given_word(word)) for word in tokens]
    return lemmatized_tokens

custom_stop_words = set(lemmatizer.lemmatize(word.lower(), get_pos_given_word(word)) for word in ENGLISH_STOP_WORDS)

lemmatized_vectorizer = CountVectorizer(tokenizer=custom_tokenizer, stop_words=list(custom_stop_words))

In [99]:
sentence = input("Enter a sentence: ")
test_fitted_model = lemmatized_vectorizer.fit([sentence])
print(ENGLISH_STOP_WORDS)
print(test_fitted_model.get_feature_names_out())
print(word_tokenize(sentence))

frozenset({'both', 'hers', 'neither', 'ourselves', 'are', 'them', 'ltd', 'now', 'nothing', 'done', 'never', 'yourselves', 'one', 'thru', 'whoever', 'therefore', 'then', 'also', 'everywhere', 'is', 'our', 'yours', 'front', 'an', 'some', 'via', 'alone', 'hereupon', 'your', 'thereby', 'two', 'next', 'beside', 'otherwise', 'latterly', 'by', 'twelve', 'the', 'a', 'seemed', 'beyond', 'still', 'was', 'be', 'sincere', 'whereupon', 'i', 'upon', 'un', 'thin', 'herself', 'they', 'serious', 'hereby', 'as', 'becoming', 'interest', 'mine', 'and', 'have', 'once', 'towards', 'himself', 'am', 'what', 'six', 'wherein', 'perhaps', 'us', 'these', 'why', 'on', 'latter', 'when', 'might', 'whether', 'few', 'amoungst', 'five', 'we', 'again', 'least', 'somewhere', 'couldnt', 'along', 'she', 'all', 'eight', 'although', 'back', 'noone', 'because', 'there', 'either', 'found', 'no', 'co', 'it', 'before', 'of', 'to', 'across', 'beforehand', 'first', 'own', 'nowhere', 'nor', 'whom', 'between', 'fifteen', 'forty', 'o

f:\Work\hands-on-nlp\venv\Lib\site-packages\sklearn\feature_extraction\text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


In [92]:
X_train_transformed = lemmatized_vectorizer.fit_transform(X_train)
X_test_transformed = lemmatized_vectorizer.transform(X_test)
lemmatized_model = MultinomialNB()
lemmatized_model.fit(X_train_transformed, Y_train)

f:\Work\hands-on-nlp\venv\Lib\site-packages\sklearn\feature_extraction\text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


MultinomialNB()

In [93]:

print(f"Training Accuracy: ${accuracy_score(Y_train, lemmatized_model.predict(X_train_transformed))}")
print(f"Testing Accuracy: ${accuracy_score(Y_test, lemmatized_model.predict(X_test_transformed))}")

Training Accuracy: $0.90535
Testing Accuracy: $0.8543


In [ ]:
review_input = input("Enter a review: ")
review_input = [review_input]
review_input = lemmatized_vectorizer.transform(review_input)
print("Positive" if lemmatized_model.predict(review_input) == 1 else "Negative") 

Negative


### Final Comment
After lemmatizing while considering POS tags, the accuracy is 0.8543 with the MultinomialNB